### PR TITLE
[circle-mpqsolver] Revisit  MPQSolver

### PR DESCRIPTION
**Pull request by stamalakhov**
This commit introduces functionality needed for `pattern` quantization.

Its correctness is tested at https://github.com/Samsung/ONE/pull/11560.

Draft: https://github.com/Samsung/ONE/pull/11560
Related: https://github.com/Samsung/ONE/issues/11543
ONE-DCO-1.0-Signed-off-by: s.malakhov <s.malakhov@partner.samsung.com>

### CODE REVIEW

File: compiler/circle-mpqsolver/src/MPQSolver.cpp [hunk 1]
```diff
@@ -16,11 +16,16 @@
...
 
+#include "pattern/PatternResolver.h"
+
 #include <luci/ImporterEx.h>
 #include <iostream>
 
 using namespace mpqsolver;
 
+using LayerParam = luci::CircleQuantizer::Options::LayerParam;
+using LayerParams = std::vector<std::shared_ptr<LayerParam>>;
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/MPQSolver.cpp:27]: You can now use
```suggestion
using LayerParams = luci::CircleQuantizer::Options::LayerParams;
```

File: compiler/circle-mpqsolver/src/MPQSolver.cpp [hunk 2]
```diff
@@ -34,6 +39,55 @@ void MPQSolver::set_save_intermediate(const std::string &save_path)
...
+    std::unique_ptr<pattern::PatternResolver> resolver;
+    switch (pattern)
+    {
+      case QuantizationPattern::Q8LayerNormWithQ16Variance:
+      {
+        resolver = std::make_unique<pattern::Q8LayerNormWithQ16VarianceResolver>();
+      }
+      break;
+      default:
+        throw std::runtime_error("ERROR: unsupported pattern to resolve");
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/MPQSolver.cpp:70]:
```suggestion
        throw std::runtime_error("Unsupported pattern to resolve");
```

File: compiler/circle-mpqsolver/src/MPQSolver.cpp [hunk 3]
```diff
@@ -34,6 +39,55 @@ void MPQSolver::set_save_intermediate(const std::string &save_path)
...
+      auto const frozen = _frozen._node_to_param.find(node_param.first);
+      if (frozen == _frozen._node_to_param.end())
+      {
+        // node was not previously defined - just set it (no ambiguity)
+        _frozen._node_to_param[node_param.first] = node_param.second;
+      }
+      else if (frozen->second.dtype != node_param.second.dtype)
+      {
+        // ambiguity (incoming description conflicts with current)
+        throw std::runtime_error("ERROR: resolved patterns contradict each other");
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/MPQSolver.cpp:85]:
```suggestion
        throw std::runtime_error("Resolved patterns contradict each other");
```

File: compiler/circle-mpqsolver/src/MPQSolver.cpp (hunk 2 above)
Comment (seanshpark) [reply compiler/circle-mpqsolver/src/MPQSolver.cpp:70]: `safemain` module will attach `ERROR: ` prefix with the messge

Comment (seanshpark) [reply compiler/circle-mpqsolver/src/MPQSolver.cpp:70]: In another PR, please change throw message not to have "ERROR: "

File: compiler/circle-mpqsolver/src/MPQSolver.cpp [hunk 4]
```diff
@@ -34,6 +39,55 @@ void MPQSolver::set_save_intermediate(const std::string &save_path)
...
+  for (auto pattern : _options._patterns)
+  {
+    std::unique_ptr<pattern::PatternResolver> resolver;
+    switch (pattern)
+    {
+      case QuantizationPattern::Q8LayerNormWithQ16Variance:
+      {
+        resolver = std::make_unique<pattern::Q8LayerNormWithQ16VarianceResolver>();
+      }
+      break;
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/MPQSolver.cpp:68]: is there any reason not to use like this?
```suggestion
        resolver = std::make_unique<pattern::Q8LayerNormWithQ16VarianceResolver>();
        break;
```

File: compiler/circle-mpqsolver/src/MPQSolver.h [hunk 5]
```diff
@@ -28,6 +28,21 @@
...
+};
+
+struct MPQOptions
+{
+  std::vector<QuantizationPattern> _patterns;
+};
+
+struct FrozenNodes
+{
+  std::map<luci::CircleNode *, luci::CircleQuantizer::Options::LayerParam> _node_to_param;
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/MPQSolver.h:43]: add include <map>

File: compiler/circle-mpqsolver/src/MPQSolver.cpp (hunk 4 above)
Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/MPQSolver.cpp:68]: > is there any reason not to use like this?

No. For now there is no reason to use `{}`.

File: compiler/circle-mpqsolver/src/MPQSolver.h (hunk 5 above)
Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/MPQSolver.h:43]: ok

File: compiler/circle-mpqsolver/src/MPQSolver.h [hunk 6]
```diff
@@ -22,12 +22,28 @@
...
 {
 
+enum class QuantizationPattern
+{
+  Q8LayerNormWithQ16Variance
+};
+
+struct MPQOptions
+{
+  std::vector<QuantizationPattern> _patterns;
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/MPQSolver.h:39]: `#include <vector>` too

Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/MPQSolver.h:39]: Ahh. Ok.

**Review by jinevening — APPROVED** [review]
LGTM

**Review by seanshpark — APPROVED** [review]
LGTM

**Pull request merged**